test/e2e/framework/types/runner.go

```diff
@@ -2,27 +2,61 @@ package types
 
 import (
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/require"
 )
 
 // A wrapper around a job, so that internal job components don't require things like *testing.T
 // and can be reused elsewhere
-type Runner struct {
+type runner struct {
 	t   *testing.T
 	Job *Job
 }
 
-func NewRunner(t *testing.T, job *Job) *Runner {
-	return &Runner{
+type runnerFunc func()
+
+func (r runnerFunc) Run() {
+	r()
+}
+
+type Runner interface {
+	Run()
+}
+
+func NewRunner(t *testing.T, job *Job) *runner {
+	return &runner{
 		t:   t,
 		Job: job,
 	}
 }
 
-func (r *Runner) Run() {
+func (r *runner) Run() {
 	if r.t.Failed() {
 		return
 	}
 	require.NoError(r.t, r.Job.Run())
 }
+
+func RunnerWithTimeout(r Runner, t time.Duration) runnerFunc {
+	timeout := make(chan bool, 1)
+	runComplete := make(chan struct{})
+
+	go func() {
+		time.Sleep(t)
+		timeout <- true
+	}()
+
+	return func() {
+		go func() {
+			r.Run()
+			runComplete <- struct{}{}
+		}()
+		select {
+		case <-timeout:
+			panic("Timeout")
```

Comment:
It's better to `testing.(*T).Fatal` in tests than outright panic. The effect is similar (it calls `runtime.Goexit()` under the hood which is panic-adjacent).